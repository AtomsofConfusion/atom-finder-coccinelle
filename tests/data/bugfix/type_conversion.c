
#include <stdio.h>

int type_conversion() {
    unsigned a = 4;
    unsigned b = 2;
    unsigned c = 3;
    unsigned d = 4;
    int e = 5;

    if (a < 2) {
        print("this is fine");
    }

    if (a < 0) {
        print("this will never happen");
    }

    if (a + 3 && b < 0 ) {
        print("this will never happen");
    }

    if (c < 0 || a + 4) {
        print("shouldn't compare unsigned to 0");
    }

    if (c + 2 || a + 4 && d < 0) {
        print("shouldn't compare unsigned to 0");
    }

    if (c < 0 && d < 0) {
        print("shouldn't compare unsigned to 0");
    }

    if (e  < 0) {
        print("this is fine, this is an integer");
    }


    return 0;
}



