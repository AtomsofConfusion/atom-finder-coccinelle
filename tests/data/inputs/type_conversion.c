#include <stdio.h>

void fun1(int x) {
  x = x + 1;
  double y = 1.5;
}

int fun2() {
  double x = 1.5;
  return x;
}

void fun3(int x, double y){
  x = y;
}

int main() {
  //Function call expression 
  double a1 = 1;
  double a2 = 1.5;
  fun1(a2);
  fun3(a1, a2);
  fun3(a2, a1);


  long d1 = 0;
  long d2 = 1;

  if (d2 > 0) {
    printf("test");
  }

  if (!d1 || d2 > 0) {
      printf("The condition was true.\n");
  } else {
      printf("The condition was false.\n");
  }



  //Binary expression
  int b1 = 1;
  double b2 = 1.5;
  int b3 = b1 + b2;
  b3 += b2;
  1 + (int) 1.5;
  (int) 1.5 + 1;
  1.5 * 2;

  //not an atom
  double c1 = 1.0;
  double c2 = c1;

  //Simple declaration
  int e1 = 1;
  double e2 = e1;
  int e3 = 1.5;
  e1 = 1.5;
  double e4 = e1 + 4;

  double abc = 2.0;
  abc = 2.5;

  //Return statement
  func2();

  //Cast expression
  double g1 = 1.5;
  int g2 = (int)g1;
  

  return 0;
}
