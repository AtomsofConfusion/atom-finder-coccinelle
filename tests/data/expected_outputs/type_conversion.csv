type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,28,2, int b3 = b1 + b2 ;
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,53,2, int g2 = ( int ) g1 ;
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,32,2,1.5 * 2
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,29,2,b3 += b2
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,42,2,e1 = 1.5
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,14,2,x = y
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,21,2,fun1 ( a2 )
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,22,2,"fun3 ( a1 , a2 )"
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,23,2,"fun3 ( a2 , a1 )"
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,10,2,return x ;
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,41,2, int e3 = 1.5 ;
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,46,2,abc = 2.5
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,30,2,1 + ( int ) 1.5
type_conversion,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/type_conversion.c,31,2,( int ) 1.5 + 1

